### PR TITLE
src: lib: Rework on singleton using mutex

### PR DESCRIPTION
**Pull request by RaulTrombin**
An update removing the unsafe calling,
Based on https://gist.github.com/mersinvald/35958fb17c14b0d827400855d8eb92e8 that was pointed on some previous PRs.

image
image

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2,23 +2,37 @@
...
-            navigator: navigator_rs::Navigator::new(),
+    fn new() -> &'static Mutex<Option<Self>> {
+        let mut st = NAVIGATOR.lock().unwrap();
+        if st.is_none() {
+            *st = Some(NavigationManager {
+                navigator: navigator_rs::Navigator::new(),
+            });
+            &NAVIGATOR
+        } else {
+            panic!("NavigationManager is already initialized")
```
Comment (patrickelectric) [inline src/lib.rs:26]: We should not abort, there is no reason for it, we could just populate NAVIGATOR content if there is nothing there, and if it's already initialized it, ignore.

File: src/lib.rs [hunk 2]
```diff
@@ -2,23 +2,37 @@
...
-    fn get_instance() -> &'static mut NavigationManager {
-        static mut SINGLETON_INSTANCE: Option<NavigationManager> = None;
-
-        // use unsafe block to ensure thread safety
-        unsafe { SINGLETON_INSTANCE.get_or_insert_with(NavigationManager::new) }
+    fn get_instance() -> &'static Mutex<Option<Self>> {
+        if NAVIGATOR.lock().unwrap().is_some() {
+            &NAVIGATOR
+        } else {
+            panic!("NavigationManager must be initialized before use")
```
Comment (patrickelectric) [inline src/lib.rs:34]: we should init on the place of a panic.

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (RaulTrombin) [reply src/lib.rs]: what about the new commit, get_instance can handle both conditions now.

File: src/lib.rs [hunk 3]
```diff
@@ -2,23 +2,27 @@
...
+lazy_static! {
+    static ref NAVIGATOR: Mutex<Option<NavigationManager>> = Mutex::new(None);
+}
 
-        // use unsafe block to ensure thread safety
-        unsafe { SINGLETON_INSTANCE.get_or_insert_with(NavigationManager::new) }
+impl NavigationManager {
+    fn get_instance() -> &'static Mutex<Option<Self>> {
+        if NAVIGATOR.lock().unwrap().is_some() {
+        } else {
```
Comment (patrickelectric) [inline src/lib.rs:19]: the else is unnecessary

**Review by patrickelectric — CHANGES_REQUESTED** [review]
Else is unnecessary

**Review by patrickelectric — APPROVED** [review]

**Pull request merged**